src/mpid/ch4/src/ch4_coll_impl.h

```diff
@@ -416,6 +419,112 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Bcast_intra_composition_gamma(void *buffer, M
     goto fn_exit;
 }
 
+/*
+ * This type of composition performs well for GPU bcast as it can utilize the direct links between
+ * the GPUs in the same node. It has four steps:
+ * 1. Root copies data to the host.
+ * 2. Inter-node bcast among the leaders on the host.
+ * 3. Leaders copy data to GPU.
+ * 4. Intra-node bcast in each node on the GPU buffer.
+*/
+MPL_STATIC_INLINE_PREFIX int MPIDI_Bcast_intra_composition_delta(void *buffer, MPI_Aint count,
+                                                                 MPI_Datatype datatype,
+                                                                 int root, MPIR_Comm * comm,
+                                                                 MPIR_Errflag_t errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int coll_ret = MPI_SUCCESS;
+    void *host_buffer = NULL;
+    void *saved_buffer = buffer;
+    MPL_pointer_attr_t attr;
+    MPI_Aint size, shift;
+
+#ifdef HAVE_ERROR_CHECKING
+    MPI_Status status;
+    MPI_Aint nbytes, recvd_size, type_size;
+#endif
+
+    if (comm->node_roots_comm == NULL && comm->rank == root) {
+        coll_ret = MPIC_Send(buffer, count, datatype, 0, MPIR_BCAST_TAG, comm->node_comm, errflag);
+        MPIR_ERR_COLL_CHECKANDCONT(coll_ret, errflag, mpi_errno);
+    }
+
+    if (comm->node_roots_comm != NULL && comm->rank != root &&
+        MPIR_Get_intranode_rank(comm, root) != -1) {
+#ifndef HAVE_ERROR_CHECKING
+        coll_ret =
+            MPIC_Recv(buffer, count, datatype, MPIR_Get_intranode_rank(comm, root), MPIR_BCAST_TAG,
+                      comm->node_comm, MPI_STATUS_IGNORE);
+        MPIR_ERR_COLL_CHECKANDCONT(coll_ret, errflag, mpi_errno);
```

Comment:
I find it difficult to sort out the branches. How about --
```
int intra_root = MPIR_Get_intranode_rank(comm, root);
if (intra_root != -1 && intra_root != 0) {
    /* root send message to local leader (node_comm rank 0) */
    if (comm->rank == root) {
        [MPIC_Send]
    } else {
        [MPIC_Recv]
    }
}
```
?

Reply:
Yes. We should clean some of the technical debt as we adding new ones.

Reply:
Fixed the code in both composition alpha and delta.